state/applicationoffers_test.go

```diff
@@ -293,6 +293,62 @@ func (s *applicationOffersSuite) TestListOffersFilterOfferNameRegexp(c *gc.C) {
 	c.Assert(offers[0], jc.DeepEquals, offer)
 }
 
+func (s *applicationOffersSuite) TestListOffersAllowedUsersOwner(c *gc.C) {
+	sd := state.NewApplicationOffers(s.State)
+	offer, owner := s.createOffer(c, "offer1", "description for offer1")
+	s.createOffer(c, "offer2", "description for offer2")
+	s.createOffer(c, "offer3", "description for offer3")
+	offers, err := sd.ListOffers(crossmodel.ApplicationOfferFilter{
+		AllowedUsers: []string{owner, "mary"},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(offers), gc.Equals, 1)
+	c.Assert(offers[0], jc.DeepEquals, offer)
+}
+
+func (s *applicationOffersSuite) TestListOffersAllowedUsers(c *gc.C) {
+	sd := state.NewApplicationOffers(s.State)
+	offer, _ := s.createOffer(c, "offer1", "description for offer1")
+	offer2, _ := s.createOffer(c, "offer2", "description for offer2")
+	s.createOffer(c, "offer3", "description for offer3")
+	s.Factory.MakeUser(c, &factory.UserParams{Name: "mary"})
+
+	mary := names.NewUserTag("mary")
+	err := s.State.CreateOfferAccess(
+		names.NewApplicationOfferTag(offer.OfferName), mary, permission.ConsumeAccess)
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.State.CreateOfferAccess(
+		names.NewApplicationOfferTag(offer2.OfferName), mary, permission.ReadAccess)
+	c.Assert(err, jc.ErrorIsNil)
+	offers, err := sd.ListOffers(crossmodel.ApplicationOfferFilter{
+		AllowedUsers: []string{"mary"},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(offers), gc.Equals, 1)
+	c.Assert(offers[0], jc.DeepEquals, offer)
+}
+
+func (s *applicationOffersSuite) TestListOffersConnectedUsers(c *gc.C) {
+	sd := state.NewApplicationOffers(s.State)
+	offer, _ := s.createOffer(c, "offer1", "description for offer1")
+	s.createOffer(c, "offer2", "description for offer2")
+	s.createOffer(c, "offer3", "description for offer3")
+	s.Factory.MakeUser(c, &factory.UserParams{Name: "mary"})
+
+	_, err := s.State.AddOfferConnection(state.AddOfferConnectionParams{
+		SourceModelUUID: testing.ModelTag.Id(),
+		Username:        "mary",
+		OfferUUID:       offer.OfferUUID,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	offers, err := sd.ListOffers(crossmodel.ApplicationOfferFilter{
+		ConnectedUsers: []string{"mary"},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(offers), gc.Equals, 1)
+	c.Assert(offers[0], jc.DeepEquals, offer)
+}
+
```

Comment:
This seems odd that you haven't allowed her to have access to the offer, but still connected to it. AddOfferConnection feels a bit like it should fail under this circumstance.

Reply:
Adding a connection is orthogonal to permission - it simply records the existence of a connection for accounting purposes. The actual connection is made via a RegisterRelation() call where permission is indeed checked.